File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    
    graph = [[-1 for _ in range(102)] for _ in range(102)]
    dp = [[-1 for _ in range(102)] for _ in range(102)]
    
    for rect in rectangle:
        startX, startY, endX, endY = map(lambda x: x * 2, rect)
        
        for i in range(startX, endX + 1):
            for j in range(startY, endY + 1):
                if startX < i < endX and startY < j < endY:
                    graph[i][j] = 0                  
                elif graph[i][j] != 0:
                    graph[i][j] = 1
    
    cX, cY, iX, iY = characterX * 2, characterY * 2, itemX * 2, itemY * 2
    
    def bfs():
        queue = deque()
        queue.append([cX, cY])
        dp[cX][cY] = 0
        
        while queue:
            curX, curY = queue.popleft()            
            
            if curX == iX and curY == iY:
                return dp[iX][iY] // 2
            
            for nextX, nextY in [[curX + 1, curY], [curX - 1, curY], [curX, curY + 1], [curX, curY - 1]]:
                if nextX < 0 or nextX >= 102 or nextY < 0 or nextY >= 102:
                    continue
                
                if graph[nextX][nextY] == 1:
                    if dp[nextX][nextY] < 0 or dp[nextX][nextY] > dp[curX][curY] + 1:
                        dp[nextX][nextY] = dp[curX][curY] + 1
                        queue.append([nextX, nextY])
    
    answer = bfs()
    return answer

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.js
function solution(rectangle, characterX, characterY, itemX, itemY) {
    let answer = 0;
    
	characterX *= 2;
	characterY *= 2;
	itemX *= 2;
	itemY *= 2;

	let map = Array.from({ length: 103 }, () => Array(101).fill(0));
	rectangle = rectangle.map((item) => item.map((x) => x * 2));
	for (let [x1, y1, x2, y2] of rectangle) {
		for (let i = x1; i <= x2; i++) {
			for (let j = y1; j <= y2; j++) {
				if (i === x1 || i === x2 || j === y1 || j === y2) {
					if (map[i][j] === 0) map[i][j] = 1;
				} else map[i][j] = 2;
			}
		}
	}

	let dx = [-1, 0, 1, 0];
	let dy = [0, -1, 0, 1];

	let queue = [[characterX, characterY, 0]];
	map[characterX][characterY] = 0;

	while (queue.length) {
		let [curX, curY, cnt] = queue.shift();

		if (curX === itemX && curY === itemY) {
            answer = cnt / 2;
            break;
        }

		for (let i = 0; i < 4; i++) {
			let nextX = curX + dx[i];
			let nextY = curY + dy[i];

			if (map[nextX][nextY] === 1) {
				queue.push([nextX, nextY, cnt + 1]);
				map[nextX][nextY] = 0;
			}
		}
	}

	return answer;
}